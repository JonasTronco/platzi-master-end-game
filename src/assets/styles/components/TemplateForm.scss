@import '../main.scss';

.template-form {
  &__title {
    margin-bottom: $space_x6;
  }

  .input {
    margin-bottom: $space_x3;
  }
}

@media screen and (max-width: 425px) {
  .input,
  .btn {
    width: 100%;
  }

  .input {
    &__label,
    &__field {
      width: 100%;
    }
  }
}
