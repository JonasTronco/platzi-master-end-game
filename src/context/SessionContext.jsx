import { useState, useEffect, createContext, useMemo, useContext } from 'react';
import Router from 'next/router';

const SessionContext = createContext();
export function SessionProvider(props) {
  const [session, setSession] = useState(null);

  useEffect(() => {
    try {
      setSession(JSON.parse(localStorage.getItem('session')));
    } catch (error) {
      Router.push('/');
    }
  }, []);

  const value = useMemo(
    () => ({
      session,
    }),
    [session],
  );

  return <SessionContext.Provider value={value} {...props} />;
}

export function useSession() {
  const context = useContext(SessionContext);
  if (!context) {
    throw new Error('No Session Context Provider');
  }
  return context;
}
