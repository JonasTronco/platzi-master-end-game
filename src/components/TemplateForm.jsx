import PropTypes from 'prop-types';
import axios from 'axios';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import { useSession } from '../context/SessionContext';
import useRequest from '../hooks/useRequest';

function ScheduleForm({ examName }) {
  const { session } = useSession();
  const { response, loading, error } = useRequest(
    session.token,
    `${process.env.NEXT_PUBLIC_API_URL}/exams?name=${examName}`,
    1,
  );

  if (loading) {
    return <div className="loader" />;
  }

  if (error) {
    return <h3>{error.message}</h3>;
  }

  return (
    <>
      {response && response.data ? (
        <Formik
          initialValues={{ value: '' }}
          validate={(values) => {
            const errors = {};
            if (!values.value) {
              errors.value = 'No fields can be empty';
            }
            return errors;
          }}
          onSubmit={async (values, { setSubmitting }) => {
            console.log(values);
            setSubmitting(false);
          }}
        >
          {({ isSubmitting }) => (
            <Form className="add-result__form">
              {response.data[0].resultTemplate.map((field) => (
                <div className="input">
                  <label className="input__label">
                    {`Field Name: ${field.fieldName}`}
                    <Field
                      className="input__field"
                      type="value"
                      name="value"
                      disabled={isSubmitting}
                    />
                  </label>
                  <ErrorMessage
                    name="value"
                    component="div"
                    className="input__error"
                  />
                </div>
              ))}
              <button className="btn" type="submit" disabled={isSubmitting}>
                Add new result
              </button>
            </Form>
          )}
        </Formik>
      ) : (
        ''
      )}
    </>
  );
}

ScheduleForm.propTypes = {
  examName: PropTypes.string.isRequired,
};

export default ScheduleForm;
