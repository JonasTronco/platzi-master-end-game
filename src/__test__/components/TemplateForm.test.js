import TemplateForm from '../../components/TemplateForm';
import { create } from 'react-test-renderer';
import { SessionProvider } from '../../context/SessionContext';
import * as nextRouter from 'next/router';
describe('TemplateForm Component', () => {
  nextRouter.useRouter = jest.fn();
  nextRouter.useRouter.mockImplementation(() => ({
    route: '/',
    query: {
      orderID: '5ecda76661b17e50c87721ab'
    }
  }));
  const obj = {
    session: {
      token:
        'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiI1ZWNkYTc2NjYxYjE3ZTUwYzg3NzIxYWIiLCJ1c2VybmFtZSI6Im1heW5lLnNuYXNlbC40MTcxIiwic2NvcGVzIjpbInNpZ25pbjphdXRoIiwic2lnbnVwOmF1dGgiLCJyZWFkOm1vdmllcyIsInJlYWQ6dXNlci1tb3ZpZXMiLCJjcmVhdGU6dXNlci1tb3ZpZXMiLCJkZWxldGU6dXNlci1tb3ZpZXMiXSwiaWF0IjoxNTkwNzI1MTIwLCJleHAiOjE1OTExNTcxMjB9.UjjAPEHWbqyPEwX6qcs2UhLvQ6rAeQIL-ME1LDnWBRo',
      user: {
        id: '5ecda76661b17e50c87721ab',
        username: 'mayne.snasel.4171',
        typeOfUser: 'Patient',
        isActive: true,
        imageURL: 'http://dummyimage.com/200x200.png/ff4444/ffffff',
        firstName: 'Mayne',
        lastName: 'Snasel',
        defaultPath: '/tests'
      }
    }
  };

  test('Should render TemplateForm component', () => {
    const component = create(
      <SessionProvider value={obj}>
        <TemplateForm examName='' />
      </SessionProvider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
