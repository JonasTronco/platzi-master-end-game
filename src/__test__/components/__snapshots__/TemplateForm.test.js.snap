// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateForm Component Should render TemplateForm component 1`] = `""`;
